### PR TITLE
Allow exiting impersonation mode

### PR DESCRIPTION
**Pull request by rluodev**
https://github.com/hackclub/hcb/assets/34289582/fb560b6a-3399-45b9-9e77-e7f7f44590a7

This PR allows admins to stop impersonating another user without logging out of their own admin user session through usage of the impersonated and impersonated_by variables of each UserSession. The above video shows updated functionality.

### CODE REVIEW

**Comment by sarthaktexas** [comment]
so coquette i love it 😻 will review in a couple hours

**Review by sarthaktexas — APPROVED** [review]
approved!

**Comment by rluodev** [comment]
@malted fix the conflict pls, i have no idea where the impersonation button is on your new redesign

File: app/controllers/users_controller.rb [hunk 1]
```diff
@@ -16,7 +16,19 @@ def impersonate
 
     impersonate_user(user)
 
-    redirect_to params[:return_to] || root_path, flash: { info: "You're now impersonating #{user.name}." }
+    redirect_to params[:return_to] || root_path, flash: { info: "You're now impersonating #{user&.name}." }
```
Comment (garyhtou) [inline app/controllers/users_controller.rb]:
```suggestion
    redirect_to params[:return_to] || root_path, flash: { info: "You're now impersonating #{user.name}." }
```

File: app/controllers/users_controller.rb [hunk 2]
```diff
@@ -16,7 +16,19 @@ def impersonate
...
+  end
+
+  def unimpersonate
+    return redirect_to root_path if !current_session&.impersonated? || !current_session&.impersonated_by&.admin?
```
Comment (garyhtou) [inline app/controllers/users_controller.rb]:
```suggestion
    return redirect_to root_path unless current_session&.impersonated?
```

File: app/controllers/users_controller.rb [hunk 3]
```diff
@@ -16,7 +16,19 @@ def impersonate
...
+
+    imp_by = current_session&.impersonated_by
```
Comment (garyhtou) [inline app/controllers/users_controller.rb]:
```suggestion
    imp_by = current_session.impersonated_by
```

File: app/controllers/users_controller.rb [hunk 4]
```diff
@@ -16,7 +16,19 @@ def impersonate
...
+
+    user = current_user
+
+    unimpersonate_user(imp_by)
+
+    redirect_to params[:return_to] || root_path, flash: { info: "Welcome back, 007. You're no longer impersonating #{user&.name}" }
```
Comment (garyhtou) [inline app/controllers/users_controller.rb]:
```suggestion
    redirect_to params[:return_to] || root_path, flash: { info: "Welcome back, 007. You're no longer impersonating #{user.name}" }
```

File: app/controllers/users_controller.rb [hunk 5]
```diff
@@ -16,7 +16,19 @@ def impersonate
...
+  end
+
+  def unimpersonate
+    return redirect_to root_path if !current_session&.impersonated? || !current_session&.impersonated_by&.admin?
+
+    imp_by = current_session&.impersonated_by
+
+    user = current_user
+
+    unimpersonate_user(imp_by)
```
Comment (garyhtou) [inline app/controllers/users_controller.rb]: Let's change this so that `unimpersonate_user` does not take in any parameters. We'll move the code above (for determining the `imp_by` into `unimpersonate_user` and handle everything in there.

File: app/views/application/_header.html.erb [hunk 6]
```diff
@@ -70,11 +70,20 @@
           'aria-label': "View transactions awaiting receipts" do %>
             <%= inline_icon "payment-docs", size: home_action_size %>
       <% end %>
-      <%= button_to logout_users_path,
-          method: :delete,
-          class: "muted tooltipped tooltipped--s",
-          'aria-label': "Sign out" do %>
-            <%= inline_icon "door-leave", size: home_action_size %>
+      <% if current_session&.impersonated? %>
```
Comment (garyhtou) [inline app/views/application/_header.html.erb]: Let's not change the sign-out button also handle existing impersonation

File: app/views/events/_organization_position_invite.html.erb
Comment (garyhtou) [inline app/views/events/_organization_position_invite.html.erb:1]: Unrelated changes?

File: app/helpers/sessions_helper.rb [hunk 7]
```diff
@@ -11,9 +11,16 @@ module SessionsHelper
   }.freeze
 
   def impersonate_user(user)
+    sign_out
     sign_in(user:, impersonate: true)
   end
 
+  def unimpersonate_user
+    curses = current_session
```
Comment (garyhtou) [inline app/helpers/sessions_helper.rb]: haha lol funny name, but i guess it makes sense

File: app/views/events/_organization_position_invite.html.erb [hunk 8]
```diff
@@ -18,9 +18,11 @@
           <div class="tooltipped tooltipped--s" aria-label="Edit user's settings">
             <%= pop_icon_to "controls", edit_user_path(position.user.slug), data: { turbo_frame: "_top" }, class: "admin-tools circle overflow-hidden" %>
           </div>
-          <div class="tooltipped tooltipped--s" aria-label="Impersonate user">
-            <%= pop_icon_to "spy", impersonate_user_path(id: position.user.id, return_to: request.url), data: { turbo_frame: "_top" }, class: "admin-tools circle overflow-hidden" %>
-          </div>
+          <% if position.user == current_user %>
```
Comment (garyhtou) [inline app/views/events/_organization_position_invite.html.erb]: hm... it seems like this makes it so that you can only impersonate yourself. is this intended behavior?

File: app/controllers/users_controller.rb [hunk 9]
```diff
@@ -220,6 +230,10 @@ def logout_session
         flash[:error] = "Error deleting the session"
         return
       end
+      if session == current_session && current_session.impersonated?
+        unimpersonate
+        return
+      end
```
Comment (garyhtou) [inline app/controllers/users_controller.rb]: Let's remove this. I want to have a clear separation between logging out and unimpersonating.

```suggestion
```

Also, I _think_ the code above (`if session.user.id != current_user.id`) will prevent this code from working correctly.

File: app/views/users/_user_session.html.erb [hunk 10]
```diff
@@ -12,8 +12,14 @@
         <span>No device is available for this session.</span>
       <% end %>
     </span>
-    <%= link_to logout_session_users_path(user: @user, id: session.id), method: :delete, class: "muted tooltipped tooltipped--w z5", 'aria-label': "Sign out of this session" do %>
-      <%= inline_icon "door-leave", size: home_action_size %>
+    <% if current_session == session && current_session&.impersonated? %>
+    <%= link_to logout_session_users_path(user: @user, id: session.id), method: :delete, class: "muted tooltipped tooltipped--w z5", 'aria-label': "Exit this impersonated session" do %>
```
Comment (garyhtou) [inline app/views/users/_user_session.html.erb]: fixes indent
```suggestion
      <%= link_to logout_session_users_path(user: @user, id: session.id), method: :delete, class: "muted tooltipped tooltipped--w z5", 'aria-label': "Exit this impersonated session" do %>
```

Comment (garyhtou) [inline app/views/users/_user_session.html.erb]: Make this POST to the unimpersonate endpoint

**Review by garyhtou — COMMENTED** [review]
couple things left!

File: app/helpers/sessions_helper.rb (hunk 7 above)
Comment (rluodev) [reply app/helpers/sessions_helper.rb]: lol

File: app/views/events/_organization_position_invite.html.erb (hunk 8 above)
Comment (rluodev) [reply app/views/events/_organization_position_invite.html.erb]: will take a look

Comment (rluodev) [reply app/views/events/_organization_position_invite.html.erb]: oop yeah that should've been an unless

File: app/views/users/_user_session.html.erb (hunk 10 above)
Comment (rluodev) [reply app/views/users/_user_session.html.erb]: all done!

**Review by garyhtou — APPROVED** [review]

**Comment by garyhtou** [comment]
Great! thanks
